The ``100-matrix_mul`` module
======================

Using ``matrix_mul``
-------------------

Import module:

    >>> matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul


	>>> matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4]])
	[[7, 10], [15, 22]]

	>>> matrix_mul([[1, 2]], [[3, 4], [5, 6]])
	[[13, 16]]

	>>> matrix_mul(12, [[4, 3]])
	Traceback (most recent call last):
		...
	ValueError: Scalar operands are not allowed, use '*' instead

	>>> matrix_mul([[4, 5], [6, 7]], 12)
	Traceback (most recent call last):
		...
	ValueError: Scalar operands are not allowed, use '*' instead

	>>> matrix_mul(12, 24)
	Traceback (most recent call last):
		...
	ValueError: Scalar operands are not allowed, use '*' instead

	>>> matrix_mul([4, 5], [[6, 7], [8, 9]])
	[64, 73]

	>>> matrix_mul([[1, 2], [3, 4]], [1, 9])
	[19, 39]

	>>> matrix_mul([1, 2], 97)
	Traceback (most recent call last):
		...
	ValueError: Scalar operands are not allowed, use '*' instead

	>>> matrix_mul([], [[4, 5]])
	Traceback (most recent call last):
		...
	ValueError: shapes (0,) and (1,2) not aligned: 0 (dim 0) != 1 (dim 0)

	>>> matrix_mul([[]], [[4, 5]])
	Traceback (most recent call last):
		...
	ValueError: shapes (1,0) and (1,2) not aligned: 0 (dim 1) != 1 (dim 0)

	>>> matrix_mul([[4, 5]], [])
	Traceback (most recent call last):
		...
	ValueError: shapes (1,2) and (0,) not aligned: 2 (dim 1) != 0 (dim 0)

	>>> matrix_mul([[4, 5]], [[]])
	Traceback (most recent call last):
		...
	ValueError: shapes (1,2) and (1,0) not aligned: 2 (dim 1) != 1 (dim 0)

	>>> matrix_mul([[]], [])
	[0.0]

	>>> matrix_mul([[4, 5], [6, None]], [[1, 2], [3, 4]])
	Traceback (most recent call last):
		...
	TypeError: Object arrays are not currently supported

	>>> matrix_mul([[4, 5], [6, 7]], [[1, 2], [None, 4]])
	Traceback (most recent call last):
		...
	TypeError: Object arrays are not currently supported

	>>> matrix_mul([[4, "foo"], [6, 7]], [[1, 2], [None, 4]])
	Traceback (most recent call last):
		...
	TypeError: Object arrays are not currently supported

	>>> matrix_mul([[1, 2, 3], [4, 5]], [[6, 7, 8], [9, 10, 100]])
	Traceback (most recent call last):
		...
	ValueError: setting an array element with a sequence.

	>>> matrix_mul([[1, 2, 3], [4, 5, 6]], [[7, 8], [9, 10, 100]])
	Traceback (most recent call last):
		...
	ValueError: setting an array element with a sequence.

	>>> matrix_mul([[1, 2, 3], [4, 5]], [[7, 8], [9, 10, 100]])
	Traceback (most recent call last):
		...
	ValueError: setting an array element with a sequence.

	>>> matrix_mul([[1, 2, 3], [4, 5, 6]], [[1, 2, 3], [4, 5, 6]])
	Traceback (most recent call last):
		...
	ValueError: shapes (2,3) and (2,3) not aligned: 3 (dim 1) != 2 (dim 0)

	>>> matrix_mul([[1, 2], [3, 4]])
	Traceback (most recent call last):
		...
	TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

	>>> matrix_mul()
	Traceback (most recent call last):
		...
	TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'
